File: sdk/javascript/examples/basic_usage.js
#!/usr/bin/env node
/**
 * Basic usage example for MirrorDNA SDK (JavaScript/Node.js).
 *
 * This example demonstrates:
 * 1. Creating a MirrorDNA client
 * 2. Creating a Master Citation
 * 3. Computing state hashes
 * 4. Creating and validating timeline events
 * 5. Checking continuity status
 *
 * No backend required - everything runs locally with files.
 */

const MirrorDNAClient = require('../mirror_dna_client');

function main() {
    console.log('='.repeat(60));
    console.log('MirrorDNA SDK - Basic Usage Example (JavaScript)');
    console.log('='.repeat(60));
    console.log();

    // 1. Initialize client
    console.log('1. Initializing MirrorDNA client...');
    const client = new MirrorDNAClient('./sdk_demo_data_js');
    console.log(`   ✓ Client initialized, data dir: ${client.dataDir}`);
    console.log();

    // 2. Create a Master Citation
    console.log('2. Creating Master Citation...');
    const citation = client.createMasterCitation(
        'agent_assistant_js_001',
        'vault_main_storage',
        '1.0.0'
    );
    console.log(`   Citation ID: ${citation.id}`);
    console.log(`   Vault ID: ${citation.vault_id}`);
    console.log(`   Checksum: ${citation.checksum.substring(0, 16)}...`);
    console.log();

    // 3. Save citation to file
    console.log('3. Saving citation to file...');
    const citationPath = client.saveCitation(citation);
    console.log(`   ✓ Saved to: ${citationPath}`);
    console.log();

    // 4. Compute state hash for arbitrary data
    console.log('4. Computing state hashes...');
    const userData = {
        name: 'Alice',
        preferences: {
            language: 'JavaScript',
            theme: 'dark'
        },
        created: '2025-11-14T10:00:00Z'
    };

    const hash1 = client.computeStateHash(userData);
    console.log(`   Hash of user data: ${hash1}`);

    // Demonstrate determinism - same data = same hash
    const hash2 = client.computeStateHash(userData);
    console.log(`   Hash (computed again): ${hash2}`);
    console.log(`   ✓ Hashes match: ${hash1 === hash2}`);
    console.log();

    // 5. Create timeline events
    console.log('5. Creating timeline events...');
    const actorId = citation.id;

    // Session start
    const event1 = client.createTimelineEvent(
        'session_start',
        actorId,
        {
            platform: 'SDK_Demo_JS',
            version: '1.0.0'
        }
    );
    console.log(`   Event 1: ${event1.event_type} (${event1.id})`);

    // Memory created
    const event2 = client.createTimelineEvent(
        'memory_created',
        actorId,
        {
            content: 'User prefers JavaScript for development',
            tier: 'long_term'
        }
    );
    console.log(`   Event 2: ${event2.event_type} (${event2.id})`);

    // Memory created
    const event3 = client.createTimelineEvent(
        'memory_created',
        actorId,
        {
            content: 'User uses dark theme',
            tier: 'long_term'
        }
    );
    console.log(`   Event 3: ${event3.event_type} (${event3.id})`);

    // Session end
    const event4 = client.createTimelineEvent(
        'session_end',
        actorId,
        {
            duration_seconds: 300,
            outcome: 'successful'
        }
    );
    console.log(`   Event 4: ${event4.event_type} (${event4.id})`);
    console.log();

    // 6. Validate timeline
    console.log('6. Validating timeline...');
    const events = [event1, event2, event3, event4];
    const validation = client.validateTimeline(events);

    console.log(`   Valid: ${validation.valid}`);
    console.log(`   Total events: ${validation.total_events}`);
    console.log(`   Event types: ${JSON.stringify(validation.event_types)}`);
    console.log(`   Unique actors: ${validation.unique_actors}`);
    console.log(`   Timespan: ${validation.timespan.first} to ${validation.timespan.last}`);
    console.log();

    // 7. Check continuity status
    console.log('7. Checking continuity status...');
    const status = client.getContinuityStatus(actorId);

    console.log(`   Identity: ${status.identity_id.substring(0, 40)}...`);
    console.log(`   Status: ${status.status}`);
    console.log(`   Total events: ${status.total_events}`);
    console.log(`   Last activity: ${status.last_activity}`);
    console.log();

    // 8. Save timeline to file
    console.log('8. Saving timeline...');
    const timelinePath = client.saveTimeline(actorId);
    console.log(`   ✓ Saved to: ${timelinePath}`);
    console.log();

    // 9. Verify checksum
    console.log('9. Verifying citation checksum...');
    const citationWithoutChecksum = Object.fromEntries(
        Object.entries(citation).filter(([k]) => k !== 'checksum')
    );
    const isValid = client.verifyChecksum(
        citationWithoutChecksum,
        citation.checksum
    );
    console.log(`   ✓ Checksum valid: ${isValid}`);
    console.log();

    // 10. Load timeline from file
    console.log('10. Loading timeline from file...');
    const loadedEvents = client.loadTimeline(timelinePath);
    console.log(`    ✓ Loaded ${loadedEvents.length} events`);
    console.log();

    console.log('='.repeat(60));
    console.log('✓ Demo completed successfully!');
    console.log('='.repeat(60));
    console.log();
    console.log('Files created:');
    console.log(`  - ${citationPath}`);
    console.log(`  - ${timelinePath}`);
    console.log();
    console.log('Continuity Status Summary:');
    console.log(`  Identity: ${status.identity_id.substring(0, 50)}...`);
    console.log(`  Status: ${status.status.toUpperCase()}`);
    console.log(`  Events: ${status.total_events}`);
    console.log(`  Valid: ${status.valid}`);
    console.log();
}

// Run the example
if (require.main === module) {
    main();
}

module.exports = main;

File: sdk/python/examples/basic_usage.py
#!/usr/bin/env python3
"""
MirrorDNA SDK - Basic Usage Example

Demonstrates how to use the MirrorDNA Python SDK for common operations:
1. Load a vault configuration
2. Compute state hash for a directory
3. Validate a timeline file
4. Get continuity status
"""

import sys
import json
from pathlib import Path

# Add parent directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent))

from mirrordna_client import MirrorDNAClient


def print_section(title: str):
    """Print section header."""
    print(f"\n{'=' * 60}")
    print(f"  {title}")
    print('=' * 60)


def main():
    """Run basic SDK examples."""
    print("MirrorDNA Python SDK - Basic Usage Example")
    print("=" * 60)

    # Initialize client
    client = MirrorDNAClient()

    # Example 1: Load vault configuration
    print_section("1. Loading Vault Configuration")

    # Use example vault from repo if it exists
    vault_path = Path(__file__).parent.parent.parent.parent / "examples" / "minimal_vault.yaml"

    if vault_path.exists():
        try:
            vault_config = client.load_vault_config(str(vault_path))
            print(f"✓ Loaded vault configuration")
            print(f"  Vault ID: {vault_config['vault_id']}")
            print(f"  Name: {vault_config['name']}")
            print(f"  Path: {vault_config['path']}")
            print(f"  Created: {vault_config['created_at']}")
        except Exception as e:
            print(f"✗ Error loading vault: {e}")
    else:
        print(f"ℹ Example vault not found at: {vault_path}")
        print("  Creating a sample vault config in memory...")

        # Create sample config
        vault_config = {
            "vault_id": "vault_demo_001",
            "name": "Demo Vault",
            "path": "./demo_data",
            "created_at": "2025-11-14T10:00:00Z"
        }
        print(f"  Created sample vault: {vault_config['vault_id']}")

    # Example 2: Compute state hash for a directory
    print_section("2. Computing Directory State Hash")

    # Use current directory as example
    test_dir = Path(__file__).parent.parent  # sdk/python directory

    try:
        state_hash = client.compute_state_hash(str(test_dir))
        print(f"✓ Computed state hash for: {test_dir}")
        print(f"  Hash: {state_hash}")
        print(f"  Length: {len(state_hash)} characters (SHA-256)")

        # Compute again to show determinism
        state_hash2 = client.compute_state_hash(str(test_dir))
        if state_hash == state_hash2:
            print(f"✓ Hash is deterministic (same result on re-computation)")
        else:
            print(f"✗ Warning: Hash changed between computations")

    except Exception as e:
        print(f"✗ Error computing state hash: {e}")

    # Example 3: Validate timeline
    print_section("3. Validating Timeline File")

    # Create a sample timeline for demonstration
    sample_timeline = {
        "timeline_id": "demo_timeline_001",
        "event_count": 2,
        "events": [
            {
                "id": "evt_001",
                "timestamp": "2025-11-14T10:00:00Z",
                "event_type": "session_start",
                "actor": "mc_demo_agent_001",
                "payload": {"platform": "Demo"}
            },
            {
                "id": "evt_002",
                "timestamp": "2025-11-14T10:05:00Z",
                "event_type": "memory_created",
                "actor": "mc_demo_agent_001",
                "payload": {"content": "User prefers Python"}
            }
        ]
    }

    # Save to temporary file
    timeline_path = Path("/tmp/demo_timeline.json")
    with open(timeline_path, 'w') as f:
        json.dump(sample_timeline, f, indent=2)

    try:
        result = client.validate_timeline(str(timeline_path))

        if result['valid']:
            print(f"✓ Timeline is valid")
            print(f"  Timeline ID: {result['timeline_id']}")
            print(f"  Event count: {result['event_count']}")
            print(f"  First event: {result.get('first_event', 'N/A')}")
            print(f"  Last event: {result.get('last_event', 'N/A')}")
        else:
            print(f"✗ Timeline validation failed")
            print(f"  Errors: {result['errors']}")

    except Exception as e:
        print(f"✗ Error validating timeline: {e}")

    # Example 4: Get continuity status
    print_section("4. Getting Continuity Status")

    try:
        status = client.get_continuity_status(
            timeline_path=str(timeline_path)
        )

        print(f"Continuity Status:")
        print(f"  Timestamp: {status['timestamp']}")
        print(f"  Timeline valid: {status['timeline_valid']}")
        print(f"  Event count: {status.get('event_count', 0)}")

        if status.get('state_hash'):
            print(f"  State hash: {status['state_hash'][:16]}...")

    except Exception as e:
        print(f"✗ Error getting continuity status: {e}")

    # Example 5: Data checksum
    print_section("5. Computing Data Checksum")

    sample_data = {
        "id": "mc_example_001",
        "version": "1.0.0",
        "vault_id": "vault_example",
        "created_at": "2025-11-14T10:00:00Z"
    }

    checksum = client.compute_data_checksum(sample_data)
    print(f"✓ Computed checksum for sample data")
    print(f"  Checksum: {checksum}")
    print(f"  Data: {json.dumps(sample_data, sort_keys=True)}")

    # Verify determinism
    checksum2 = client.compute_data_checksum(sample_data)
    if checksum == checksum2:
        print(f"✓ Checksum is deterministic")

    # Summary
    print_section("Summary")
    print("""
The MirrorDNA SDK provides simple tools for:
- Loading and validating vault configurations
- Computing deterministic state hashes
- Validating timeline event logs
- Tracking continuity status

All operations are local-only and use standard Python libraries.

Next steps:
- Explore the full protocol in src/mirrordna/
- Check out examples/ for more advanced usage
- Read docs/ for protocol specifications
    """)

    print("✓ Example completed successfully!\n")


if __name__ == "__main__":
    main()
